source/develop/release-management/features/storage/cd-block-storage.md

```diff
@@ -0,0 +1,228 @@
+      <target dev='sdc' bus='sata'/>
+      <readonly/>
+      <alias name='ua-79287c04-4eea-4db7-a376-99a9f85ad0ed'/>
+      <address type='drive' controller='0' bus='0' target='0' unit='2'/>
+    </disk>
+```
+
+After loading CD it chages e.g. to
+
+```
+    <disk type='block' device='cdrom'>
+      <driver name='qemu' type='raw' error_policy='report'/>
+      <source dev='/rhev/data-center/mnt/blockSD/88252cf6-381e-48f0-8795-a294a32c7149/images/89f05c7d-b961-4935-993f-514499024515/626a493f-5214-4337-b580-96a1ce702c2a' index='2'>
+        <seclabel model='dac' relabel='no'/>
+      </source>
+      <backingStore/>
+      <target dev='sdc' bus='sata'/>
+      <readonly/>
+      <alias name='ua-79287c04-4eea-4db7-a376-99a9f85ad0ed'/>
+      <address type='drive' controller='0' bus='0' target='0' unit='2'/>
+    </disk>
+```
+
+so that it contains path to the ISO.
+
+Besides this libvirt XML, oVirt adds its own XML for each disk to be able to identify
+the volume by PDIV (pool, domain, image volume) coordinates and includes other useful info like lease path. e.g.
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc">
+        <ovirt-vm:domainID>88252cf6-381e-48f0-8795-a294a32c7149</ovirt-vm:domainID>
+        <ovirt-vm:imageID>89f05c7d-b961-4935-993f-514499024515</ovirt-vm:imageID>
+        <ovirt-vm:poolID>13345997-b94f-42dd-b8ef-a1392f65cebf</ovirt-vm:poolID>
+        <ovirt-vm:volumeID>626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:volumeID>
+        <ovirt-vm:volumeChain>
+            <ovirt-vm:volumeChainNode>
+                <ovirt-vm:domainID>88252cf6-381e-48f0-8795-a294a32c7149</ovirt-vm:domainID>
+                <ovirt-vm:imageID>89f05c7d-b961-4935-993f-514499024515</ovirt-vm:imageID>
+                <ovirt-vm:leaseOffset type="int">105906176</ovirt-vm:leaseOffset>
+                <ovirt-vm:leasePath>/dev/88252cf6-381e-48f0-8795-a294a32c7149/leases</ovirt-vm:leasePath>
+                <ovirt-vm:path>/rhev/data-center/mnt/blockSD/88252cf6-381e-48f0-8795-a294a32c7149/images/89f05c7d-b961-4935-993f-514499024515/626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:path>
+                <ovirt-vm:volumeID>626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:volumeID>
+            </ovirt-vm:volumeChainNode>
+        </ovirt-vm:volumeChain>
+    </ovirt-vm:device>
+```
+
+When CD is not loaded, this element is still present (as CD-ROM is always present), but is empty, e.g.:
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc"/>
+```
+
+### Bugs
+
+- [Bug 1589763: Error changing CD for a running VM when ISO image is on a block domain](https://bugzilla.redhat.com/1589763)
+- [Bug 1868643: Volume not deactivated after ejecting CDROM](https://bugzilla.redhat.com/1868643)
+
+
+### Functionality added with this feature
+
+This feature adds new APIs to vdsm for loading, ejecting and changing CD.
+With these functions, all aforementioned functionality will work properly.
+
+#### Vdsm
+
+- Add VM.insertCD API
+- Add VM.ejectCD API
```

Comment:
few years ago we started changing the virt-related parts in the engine<->vdsm api to be closer to that of libvirt
we changed create-vm but also attach/detach nic (see HotPlugNicVDSCommand and HotUnplugNicVDSCommand) and didn't get to change-cd
the nice parts in this approach:
1. VDSM doesn't need to generate the XML
2. we can reuse the same code that generates the XML for create-VM
3. by adding the xml to the dictionary of the existing verb we don't need new cluster-level or host capability or anything like that - new VDSM respect the XML if it is provided and otherwise use the other provided parameters. old VDSM ignores the XML.

can we do something similar in this case?

Reply:
+1 
it should be as close to it as possible. That's our "API" to modify VMs.
We don't have the path to pass to libvirt here as it's going to be made up by prepareVolume, but still I don't see a reason to differ here.

Also, empty and inserted CD are represented in libvirt by an "empty path" and "some path". We can keep it the same here and have some PDIV vs empty.

Reply:
Also, can you clarify if you mean to keep the API synchronous? I.e. do all those actions, activation, etc, are completed before the API call returns.

Reply:
@michalskrivanek about the 'prepared' path, the engine already knows how to predict it on create-vm (see [LibvirtVmXmlBuilder#writeDiskSource](https://github.com/oVirt/ovirt-engine/blob/8203fad23f1de20c6c800a2e4352464bb230ab3d/backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/LibvirtVmXmlBuilder.java#L2116)) - we can call the same code

Reply:
oh, right, thank you, i forgot that we actually finished that

Reply:
I'm not sure I follow. vdsm generate XML just for vdsm to have PDIV stored for the image. The issue is that vdsm activate the volume only when engine specifies volume using PDIV. If engine sends just path, vdsm assume it exists. So you propose to keep everything and just activate the volume in case of engine sends path? (IIRC we excluded this approach in previous discussions).

As for the API, yet, it's meant to be synchronous.

Reply:
vdsm doesn't generate xml. engine generates all the xml except for some parts which we try to minimize. The reesulting path where that volume is mounted is defined by engine already ([here](https://github.com/oVirt/ovirt-engine/blob/8203fad23f1de20c6c800a2e4352464bb230ab3d/backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/VmInfoBuildUtils.java#L1176))
I don't think Arik meant to send only the path. You still need additional parameters in metadata, that's fine. It's just that the path doesn't have to be replaced/generated in vdsm.
So in mind the API should be similar to other hotplugs - a verb like "changeCD" and an xml updating the <disk>'s path and corresponding metadata changes (PDIV, other properties, whatever you need). Both ways for insert and eject
API - ack. that makes it simpler

Reply:
so you mean replace e.g. [this](https://github.com/oVirt/vdsm/blob/master/lib/vdsm/virt/vm.py#L4375) where we change path in VM XML by using XML will be sent by engine and on vdsm just make sure volume defined by sent PDIV is activated?

Reply:
to be concrete, here's an example of xml that the engine generates for hotplug-nic:
```xml
<?xml version="1.0" encoding="UTF-8"?><hotplug>
  <devices>
    <interface type="bridge">
      <model type="virtio"/>
      <link state="up"/>
      <source bridge="ovirtmgmt"/>
      <alias name="ua-15f4ba44-8e3b-4bbe-817a-f118cd2c88c7"/>
      <mac address="56:6f:a7:1a:00:0a"/>
      <mtu size="1500"/>
      <filterref filter="vdsm-no-mac-spoofing"/>
      <bandwidth/>
    </interface>
  </devices>
  <metadata xmlns:ovirt-vm="http://ovirt.org/vm/1.0">
    <ovirt-vm:vm>
      <ovirt-vm:device mac_address="56:6f:a7:1a:00:0a">
        <ovirt-vm:custom/>
      </ovirt-vm:device>
    </ovirt-vm:vm>
  </metadata>
</hotplug>
```
it's ok to have both `device` and `metadata` sections.
in case of change-cd we can set the PDIV details within the metadata section like we do in create-vm, for instance:
```xml
<ovirt-vm:device devtype="disk" name="vda">
   <ovirt-vm:poolID>542d569a-74a1-4847-9c14-08c20182e7f3</ovirt-vm:poolID>
   <ovirt-vm:volumeID>1d82ee44-884b-4681-9fa3-5dcf515c1e67</ovirt-vm:volumeID>
   <ovirt-vm:imageID>152440e6-f383-466f-9724-73b3877c4e2c</ovirt-vm:imageID>
   <ovirt-vm:domainID>dde147c1-60ec-4e99-9a6e-6a709975b65b</ovirt-vm:domainID>
</ovirt-vm:device>
```

Reply:
I don't see how sending xml from engine helps. This just complicates everything and we don't do this for any storage API.

The main issue here is supporting block storage for cdrom.
- We need to always keep PDIV information in vm metadata, this is done now only when starting VM with a cdrom on block device
- We need to tear down the volume after ejecting a CD
- We need to prepare a volume before inserting a CD
- For handling PDIV, vdsm need to accept dict with the uuids. Accepting xml with the uuids does no make sense.
- We need new APIs to make the flow easier to manage
- For new APIs we need new cluster version to make it sane to manage on engine side
- Enigne cannot predict block volume path, what it does today is wrong and unsupported by vdsm. Volume path is internal implementation detail and any vdsm client that "predict" this is going to fail sooner or later. I warned about this many times.

Reply:
Sorry for commenting on this without knowing all the details and please correct me as needed. A general approach since introducing Engine XML is using XML for device handling (although it must be amended or completed on the host side in some cases). Ideally, we should be consistent and use XML wherever possible, which is unfortunately not true yet. If we can provide a CD at VM start (I guess we need that at least when booting from the CD) in the form of XML (including metadata), it makes sense to provide it in the same form when changing the CD. Sure, extracting the info from the XML is a "useless" extra step on the Vdsm side (not sure about the Engine side) but we do that on VM startup anyway. When introducing a new API, which we don't want to change in the future, we should consider the general aim of Engine XML API consistency, even if it means a bit of extra work now. It may return the investment later, especially if we ever manage to cleanup storage device handling and make another new shiny motorbike from the old rusty one. On the Vdsm side, It's about replacing dictionary queries with XPath queries and replacing a complete XML output processing with selective DOM changes. Which shouldn't be *that* big deal. Let's consider which of the two options (XML x dictionary) will make future maintenance easier from the global point of view and whether using XML could help start fixing the current storage device handling mess or would make it even worse.

Reply:
yeah, and from the engine side it enables us to reuse the code that creates the cd-rom device on create-vm also for creating the device with the different CD - if one ever changes the properties of the cd-rom device on create-vm, it will auto-magically apply in the case of change-cd without any further changes in VDSM.

at the more conceptual level I wrote back then [a post on why we took that direction](https://ahadas.com/engine-xml/). it may look like an additional overhead now but as we're about to change the VDSM<->engine API, we can take this opportunity to better align it with other verbs in VDSM in which the engine XML is used.

Reply:
It seems there's no agreement on this as @nirs is strongly against it. Nevertheless we at least agree that we will add two new APIs into vdsm. We can discuss further once the initial patches for API will be post, what should it accecpt (xml vs. dict)

Reply:
I'm not sure there is even agreement on 2 new APIs. @ahadas brought up (ok, maybe only in our conversation and not here) a point that we can easily use the same API verb just extended the dictionary with "xml" attribute, carrying the new xml representation with pdiv.

Reply:
> I don't see how sending xml from engine helps. This just complicates everything and we don't do this for any storage API.

we do that vdsm for every aspect of VM configuration since 4.2. Yes, storage code is behind because the storage team decided to ignore that change. It doesn't complicate anything, it gets the storage code aligned with the rest of vdsm.

> The main issue here is supporting block storage for cdrom.

yes

>     * We need to always keep PDIV information in vm metadata, this is done now only when starting VM with a cdrom on block device

yes, no argument there. It should be up to date

>     * We need to tear down the volume after ejecting a CD

after ejecting from the last guest using the same CD

>     * We need to prepare a volume before inserting a CD

yes (supposing doing it twice in case it's already used by other guest is harmless) 

>     * For handling PDIV, vdsm need to accept dict with the uuids. Accepting xml with the uuids does no make sense.

xml makes it aligned with all the other VM APIs

>     * We need new APIs to make the flow easier to manage

there's no complexity in API on vdsm side, neither on engine side. Plus if we can easily keep using the same API I would prefer to keep compatibility and the code simpler

>     * For new APIs we need new cluster version to make it sane to manage on engine side

this is just a complexity introduced by new API. No new API == no need to do this.

>     * Enigne cannot predict block volume path, what it does today is wrong and unsupported by vdsm. Volume path is internal implementation detail and any vdsm client that "predict" this is going to fail sooner or later. I warned about this many times.

Engine is the only meaningful vdsm client, so for better or worse this is what is supported.
We can use placeholders to replace it with calculated path from metadata (like we use for serial number). If you'd like to close the xml refactoring finishing the storage subsystem transition this would be a good task to do...

Reply:
I agree, it looks like a good plan to me.

Reply:
+1

Reply:
LGTM

Reply:
> After some more offline discussion, can we agree on following @nirs @michalskrivanek @ahadas @mz-pdm ?
> 
> * engine will send XML

Maybe, we will discuss these details when we work on the implementation.

> * virt part of vdsm will accept XML
> * virt part of vdsm will create a dict with PDIV from XML and will pass it to storage part
> * if there is no XML in the request, path will be used (i.e. we will fall back to old behavior)
> * fall back to old behavior described in previous item means we don't need new vdsm API and can reuse the old one without breaking backward compatibility

No, we are not going to do these kinds of hacks. We need clean API, no hacks.

> If you agree on these items, I'll update the doc appropriately.

I strongly disagree with this plan.

Reply:
if there is a disagreement even at this phase then it should be resolved, not dismissed...
it won't help to get to a patch stage and get a -2 there.